File: DesignPatterns/src/C++/002-GOFPatterns/001-CreationalPatterns/001-Builder/AlertDialogNestedClass/AlertDialog.cpp
#include <iostream>
#include <string>
#include "AlertDialog.h"

using namespace std;

AlertDialog::Builder  &AlertDialog::Builder::setTitle(const std::string &title)
{
	m_up->m_title = title;

	return *this;
}
AlertDialog::Builder  &AlertDialog::Builder::setText(const std::string &text)
{
	m_up->m_text = text;

	return *this;
}
AlertDialog::Builder  &AlertDialog::Builder::setIcon(IconType icon)
{
	m_up->m_icon = icon;

	return *this;
}

AlertDialog::Builder  &AlertDialog::Builder::setButton(ButtonType button)
{
	m_up->m_button = button;

	return *this;
}


std::unique_ptr<AlertDialog> AlertDialog::Builder::build()
{
	//...
	return std::move(m_up);	
}

void AlertDialog::Show()
{
	string icons[]{ "QUESTION", "WARNING", "CRITICAL" };
	string buttons[]{ "YESNO", "YESNOCANCEL", "OK" };

	cout << "Title:" << m_title << endl;
	cout << "Text:" << m_text << endl;
	cout << "Icon:" << icons[m_icon] << endl;
	cout << "Button:" << buttons[m_button] << endl;
}

File: src/DesignPatterns/src/Java/002-GOF-Patterns/001-CreationalPatterns/004-Builder/AlertDialog/src/org/csystem/app/AlertDialog.java
package org.csystem.app;

public class AlertDialog {
    private String m_title;
    private String m_message;
    private AlertDialogIcon m_icon;
    private AlertDialogButton m_button;

    private AlertDialog()
    {
        m_title = "";
        m_message = "";
        m_icon = AlertDialogIcon.NONE;
        m_button = AlertDialogButton.OK;
    }

    public static class Builder {
        private final AlertDialog m_dlg = new AlertDialog();

        public Builder setTitle(String title)
        {
            m_dlg.m_title = title;

            return this;
        }

        public Builder setMessage(String message)
        {
            m_dlg.m_message = message;

            return this;
        }

        public Builder setIcon(AlertDialogIcon icon)
        {
            m_dlg.m_icon = icon;

            return this;
        }

        public Builder setButton(AlertDialogButton button)
        {
            m_dlg.m_button = button;

            return this;
        }

        public AlertDialog build()
        {
            return m_dlg;
        }
    }

    public void show()
    {
        System.out.printf("%s", m_title.isEmpty() ? "" : "Title:" + m_title + "\n");
        System.out.printf("%s", m_message.isEmpty() ? "" : "Message:" + m_message  + "\n");
        System.out.printf("%s", m_icon == AlertDialogIcon.NONE ? "" : "Icon:" + m_icon  + "\n");
        System.out.printf("%s", m_button == AlertDialogButton.NONE ? "" : "Button:" + m_button + "\n");
    }

}
